lib/src/phoenix_presence.dart

```diff
@@ -0,0 +1,175 @@
+          joins[key] = clone(newPresence);
+          joins[key]['metas'] = List<Map<String, dynamic>>.from(joinedMetas);
+        }
+
+        if (leftMetas.length > 0) {
+          leaves[key] = clone(currentPresence);
+          leaves[key]['metas'] = List<Map<String, dynamic>>.from(leftMetas);
+        }
+      } else {
+        joins[key] = newPresence;
+      }
+    });
+
+    return syncDiff(state, {'joins': joins, 'leaves': leaves}, onJoinCallback, onLeaveCallback);
+  }
+
+  static syncDiff(currentState, diffs, onJoinCallback, onLeaveCallback){
+    var joins = diffs['joins'] ?? {};
+    var leaves = diffs['leaves'] ?? {};
+    var state = clone(currentState);
+    if (onJoinCallback == null) { onJoinCallback = _noop; }
+    if (onLeaveCallback == null) { onLeaveCallback = _noop; }
+
+    joins.forEach((key, newPresence) {
+      var currentPresence = state[key];
+      state[key] = newPresence;
+      if (currentPresence != null) {
+        var joinedRefs = state[key]['metas']
+          .where((Map<String, dynamic> meta) => meta.containsKey('phx_ref'))
+          .map((Map<String, dynamic> meta) => meta['phx_ref']);
+
+        var curMetas = List<Map<String, dynamic>>.from(currentPresence['metas'].where((meta) => !joinedRefs.contains(meta['phx_ref'])));
+        state[key]['metas'].insertAll(0, curMetas);
+      } 
+      onJoinCallback(key, currentPresence, newPresence);
+    });
+
+    leaves.forEach((key, leftPresence) {
+      var currentPresence = state[key];
+      if (currentPresence == null) {
+        return;
+      }
+
+      var refsToRemove = List.from(leftPresence['metas'])
+        .where((meta) => meta.containsKey('phx_ref'))
+        .map((meta) => meta['phx_ref']);
+
+      var currentPresenceMetas = List.from(currentPresence['metas']);
+      currentPresenceMetas.removeWhere((meta) => refsToRemove.contains(meta['phx_ref']));
+      currentPresence['metas'] = currentPresenceMetas;
+
+      onLeaveCallback(key, currentPresence, leftPresence);
+
+      if (currentPresence['metas'] == null || currentPresence['metas'].length == 0) {
+        state.remove(key);
+      }
+    });
+
+    return state; 
+  }
+
+  static _list(presences, chooser){
+    if (chooser == null) {
+      chooser = (key, pres) => pres;
+    }
+
+    return List.from(presences.entries)
+      .map((entry) => chooser(entry.key, entry.value));
+  }
+
+  static clone(object){
+    return jsonDecode(jsonEncode(object));
+  }
+}
+
+class PhoenixPresenceEvents {
+  static const presenceState = "presence_state";
+  static const presenceDiff = "presence_diff";
+}
+
+class PresenceEvents {
+  String state;
+  String diff;
+  PresenceEvents(this.state, this.diff);
+}
+
+class _PresenceCallers {
+  Function(dynamic key, dynamic currentPresence, dynamic newPresence) onJoin;
+  Function(dynamic key, dynamic currentPresence, dynamic newPresence) onLeave;
+  Function() onSync;
+  _PresenceCallers(this.onJoin, this.onLeave, this.onSync);
```

Comment:
As per suggestion above, maybe this is better with named parameters as it's clearer at the invocation site

```suggestion
  _PresenceCallers({this.onJoin, this.onLeave, this.onSync});
```